src/psyclone/psyir/tools/call_tree_utils.py

```diff
@@ -0,0 +1,415 @@
+            PSyIR operators lbound, ubound, or size will be reported as
+            'read'. Otherwise, these accesses will be ignored.
+
+        '''
+        # Collect the information about all variables used:
+        if not variables_info:
+            variables_info = VariablesAccessInfo(node_list, options=options)
+
+        for signature in variables_info.all_signatures:
+            if variables_info.is_written(signature):
+                read_write_info.add_write(signature)
+
+    # -------------------------------------------------------------------------
+    def get_in_out_parameters(self, node_list, options=None):
+        '''Returns a ReadWriteInfo object that contains all variables that are
+        input and output parameters to the specified node list. This function
+        calls `get_input_parameter` and `get_output_parameter`, but avoids the
+        repeated computation of the variable usage.
+
+        :param node_list: list of PSyIR nodes to be analysed.
+        :type node_list: List[:py:class:`psyclone.psyir.nodes.Node`]
+        :param options: a dictionary with options for the CallTreeUtils
+            which will also be used when creating the VariablesAccessInfo
+            instance if required.
+        :type options: Optional[Dict[str, Any]]
+        :param Any options["COLLECT-ARRAY-SHAPE-READS"]: if this option is
+            set to a True value, arrays used as first parameter to the
+            PSyIR operators lbound, ubound, or size will be reported as
+            'read'. Otherwise, these accesses will be ignored.
+
+        :returns: a ReadWriteInfo object with the information about input-
+            and output parameters.
+        :rtype: :py:class:`psyclone.psyir.tools.ReadWriteInfo`
+
+        '''
+        variables_info = VariablesAccessInfo(node_list, options=options)
+        read_write_info = ReadWriteInfo()
+        self.get_input_parameters(read_write_info, node_list, variables_info)
+        self.get_output_parameters(read_write_info, node_list, variables_info)
+        return read_write_info
+
+    # -------------------------------------------------------------------------
+    def get_non_local_read_write_info(self, node_list, read_write_info):
+        '''Returns the information about non-local variables that are read
+        or written.
+
+        '''
+
+        # First collect all non-local symbols from the kernels called. They
+        # are collected in the todo list. This list will initially contain
+        # unknown accesses, since at this stage we cannot always differentiate
+        # between function calls and array accesses. In the resolve step
+        # that is following the corresponding modules will be queried and
+        # the right accesses (functions or variables) will be used.
+        todo = []
+        mod_manager = ModuleManager.get()
+        for node in node_list:
+            for kernel in node.walk(Kern):
```

Comment:
I am not sure, do we need to distinguish between language-level and high-level PSyIR? In some case (elsehwere, can't remember where now, kernel extraction likely), I handle kernels and calls in the same branch. So if this can be maintained (I would expect that we need to adjust things a bit when we use nemo with the module manager), I would suggest to leave this for now. And then add a separate 'renaming/moving' PR, since we also agreed that the ModuleManager etc should be elsewhere, not in parse. Let's sort the proper location out during this PR, which should make the history a lot easier to read.

Reply:
I've added a new issue for bringing the LFRic functionality back into gocean and 'generic' in #2494.